### ISSUE TITLE
Unauthorized access retrieving token after username/password login

### ISSUE THREAD

**Issue by bpair** [influxdata/influxdb-client-java#165]
After setting up influxdb 2.0 with a default user, organization, and password. Use the java client and 

```
adminClientOptions = InfluxDBClientOptions.builder()
                .url(influxURL)
                .org(orgName)
                .authenticate(basicAuthUserName, basicAuthAppPass.toCharArray())
                .build();

//Start with basic auth
adminInfluxDBClient = InfluxDBClientFactory.create(adminClientOptions);
authorizationsApi = adminInfluxDBClient.getAuthorizationsApi();

//Get Authorization
List<Authorization> authorizationsByUserName = authorizationsApi.findAuthorizationsByUserName(basicAuthUserName);
```
A UnauthorizedException is thrown. I would expect to be able to retrieve the token and use that token for creating a new client for Writes.

The reason I want to use the basic authentication first is because I do not know the token. Imagine distributing thousands of edge devices that startup and install influx and create the default user, organization and bucket. The token is different on every edge device.  However to create a long living client for writes it is preferred to use a token so I am following a two step process.

Maybe I am following the wrong pattern.

thanks.

**Comment by bednar** [comment]
I think the best approach for your use case is:

1. Create the default user, organization, bucket and token via external tool. Something like:
```console
curl -i -X POST http://localhost:9999/api/v2/setup -H 'accept: application/json' \
    -d '{
            "username": "my-user",
            "password": "my-password",
            "org": "my-org",
            "bucket": "my-bucket",
            "token": "my-token"
        }'
```
2. Bootstrap java client with configuration file - `influx2.properties`: 
```properties
influx2.url=http://localhost:8086
influx2.org=my-org
influx2.bucket=my-bucket
influx2.token=my-token
influx2.logLevel=BODY
influx2.readTimeout=5s
influx2.writeTimeout=10s
influx2.connectTimeout=5s
```
```java
InfluxDBClientOptions options = InfluxDBClientOptions.builder()
    .loadProperties()
    .build();

InfluxDBClient influxDBClient = InfluxDBClientFactory.create(options);
```

This will be easy to manage and your java code on edge device will much more easy to maintenance. 

Regards

**Comment by bpair** [comment]
Thanks for the quick response and I am sure your strategy will work but I think it is strange that the java client does not operate the same way I have been using the API. I have java code on each edge device that does the influx setup and also the writes and queries.

It would be cumbersome to use curl just to get access to the generated token and store it in a properties file. Is there a reason the java client acts differently? Maybe I am using the API incorrectly?

**Comment by bednar** [comment]
It was just a suggestion... that scenario we use in our in-house project. 

Is your user an admin?

I just tried this code:

```java
InfluxDBClient influxDBClient = InfluxDBClientFactory
        .create(influxDB_URL, "my-user", "my-password".toCharArray());

List<Authorization> authorizationsApi = influxDBClient
        .getAuthorizationsApi()
        .findAuthorizationsByUserName("my-user");

for (Authorization authorization : authorizationsApi) {
    System.out.println("token = " + authorization.getToken());
}
```

the InfluxDB was onboarded by:

```console
curl -i -X POST http://localhost:9999/api/v2/setup -H 'accept: application/json' \
    -d '{
            "username": "my-user",
            "password": "my-password",
            "org": "my-org",
            "bucket": "my-bucket",
            "token": "my-token"
        }'
```

and everything work correctly.

**Comment by bpair** [comment]
That is the same as my code. Maybe something else is wrong. I will test some more.

**Comment by bednar** [comment]
Which version of InfluxDB do you use? Are they `basicAuthUserName` and `basicAuthAppPass` credentials for `admin`?

**Comment by bpair** [comment]
Influxdb 2rc. They are the username/password that were used in onboarding.

**Comment by bpair** [comment]
I figured out where the bug is. It is not that Authorizations cannot be after logging in with a username password, the issue is how InfluxDBClientOptions handles the URL. It appears to be messing up the URL for the AuthorizationsAPI. When I login like this:

```
InfluxDBClientOptions adminClientOptions = InfluxDBClientOptions.builder()
                .url(influxURL)
                .org(orgName)
                .authenticate(basicAuthUserName, basicAuthAppPass.toCharArray())
                .build();
 
InfluxDBClient adminInfluxDBClient = InfluxDBClientFactory.create(adminClientOptions);
AuthorizationsApi authorizationsApi = adminInfluxDBClient.getAuthorizationsApi();
List<Authorization> authorizationsByUserName =
                    authorizationsApi.findAuthorizationsByUserName(basicAuthUserName);
```

I get an UnauthorizedException because I believe the URL is incorrect. If I do not use the InfluxDBClientOptions and instead use:

```
InfluxDBClientFactory.create(influxURL, basicAuthUserName, basicAuthAppPass.toCharArray());
```

the code works as expected.

**Comment by bednar** [comment]
Hi @bpair,

The `InfluxDBClientFactory.create` uses `InfluxDBClientOptions` see - https://github.com/influxdata/influxdb-client-java/blob/aee00f866628733b56030053893fee6f93177780/client/src/main/java/com/influxdb/client/InfluxDBClientFactory.java#L147

Please, could you enable http logging by `influxDBClient.setLogLevel(LogLevel.BODY);` and upload stacktrace and full response from InfluxDB?

**Comment by bednar** [comment]
Hi @faltimiras,

> Same issue here, with the 3.4 version...
> ```
> InfluxDBClientOptions adminClientOptions = InfluxDBClientOptions.builder()
> 				.url("http://localhost:8086")
> 				.org("my-org")
> 				.authenticate("my-user", "my-password".toCharArray())
> 				.build();
> 
> 		InfluxDBClient adminInfluxDBClient = InfluxDBClientFactory.create(adminClientOptions);
> 		adminInfluxDBClient.setLogLevel(LogLevel.BODY);
> 		AuthorizationsApi authorizationsApi = adminInfluxDBClient.getAuthorizationsApi();
> 		List<Authorization> authorizationsByUserName =
> 				authorizationsApi.findAuthorizationsByUserName("my-user");
> ```

I tried your code with the latest client version and the latest InfluxDB version and everything works fine. Can you please try your code with a clean installation of InfluxDB? For example with Docker:

```bash
docker run \
       --detach \
       --name influxdb_v2 \
       --network influx_network \
       --publish 8086:8086 \
       influxdb:latest
```

> With 4.1 i just got java.lang.NoSuchMethodError: 'okhttp3.RequestBody okhttp3.RequestBody.create(java.lang.String, okhttp3.MediaType)'
at com.influxdb.client.internal.AuthenticateInterceptor.initToken(AuthenticateInterceptor.java:122)
at com.influxdb.client.internal.AbstractInfluxDBClient.(AbstractInfluxDBClient.java:104)
at com.influxdb.client.internal.AbstractInfluxDBClient.(AbstractInfluxDBClient.java:78)
at com.influxdb.client.internal.InfluxDBClientImpl.(InfluxDBClientImpl.java:95)
at com.influxdb.client.InfluxDBClientFactory.create(InfluxDBClientFactory.java:204)

It looks like version conflict in `okhttp3` library. Can you share your dependency tree? For `maven` you can use something like: `mvn dependency:tree -U`.

Regards

**Comment by faltimiras** [comment]
Hi Bednar,

I just create a plain project and it works fine, you are right, it must be a problem somewhere in my project.

I have a really something weird in my project...
Where it doen't work:
faltimiras@faltimiras:~/dev/src/dtz/zentral/metrics$ mvn dependency:tree  | grep okhttp -B10
[INFO] |  \- org.latencyutils:LatencyUtils:jar:2.0.3:runtime
[INFO] +- io.micrometer:micrometer-registry-influx:jar:1.8.1:compile
[INFO] |  \- org.slf4j:slf4j-api:jar:1.7.32:compile
[INFO] +- **com.influxdb:influxdb-client-java:jar:4.1.0:compile**
[INFO] |  +- com.influxdb:influxdb-client-core:jar:4.1.0:compile
[INFO] |  |  +- com.influxdb:influxdb-client-utils:jar:4.1.0:compile
[INFO] |  |  +- com.squareup.okio:okio:jar:2.6.0:compile
[INFO] |  |  |  +- org.jetbrains.kotlin:kotlin-stdlib:jar:1.6.10:compile
[INFO] |  |  |  |  \- org.jetbrains:annotations:jar:13.0:compile
[INFO] |  |  |  \- org.jetbrains.kotlin:kotlin-stdlib-common:jar:1.6.10:compile
[INFO] |  |  +- **com.squareup.okhttp3:okhttp:jar:3.14.9:compile**
[INFO] |  |  +- com.squareup.retrofit2:retrofit:jar:2.9.0:compile
[INFO] |  |  +- com.squareup.okhttp3:logging-interceptor:jar:3.14.9:compile

Clean project:
faltimiras@faltimiras:~/dev/src/dtz/zentral/testmvn$ mvn dependency:tree 
[INFO] 
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ testmvn ---
[INFO] org.example:testmvn:jar:1.0
[INFO] \- **com.influxdb:influxdb-client-java:jar:4.1.0:compile**
[INFO]    +- com.influxdb:influxdb-client-core:jar:4.1.0:compile
[INFO]    |  +- com.influxdb:influxdb-client-utils:jar:4.1.0:compile
[INFO]    |  +- com.squareup.okio:okio:jar:2.6.0:compile
[INFO]    |  |  +- org.jetbrains.kotlin:kotlin-stdlib:jar:1.3.70:compile
[INFO]    |  |  |  \- org.jetbrains:annotations:jar:13.0:compile
[INFO]    |  |  \- org.jetbrains.kotlin:kotlin-stdlib-common:jar:1.3.70:compile
[INFO]    |  +- **com.squareup.okhttp3:okhttp:jar:4.7.2:compile**
[INFO]    |  +- com.squareup.retrofit2:retrofit:jar:2.9.0:compile
[INFO]    |  +- com.squareup.okhttp3:logging-interceptor:jar:4.7.2:compile

Thanks for your help and time, but looks like i have a gremlin in my laptop :crying_cat_face:

**Comment by faltimiras** [comment]
After digging with grep I spot the problem, my real project uses spring-boot (last version 2.6.2) which uses defines in the main pom:<okhttp3.version>3.14.9</okhttp3.version>

Then dependency:tree lies saying that is influx clien who is adding this dependency

The fix is force to use the 4.7.2 version adding, those dependencies in the pom

<dependency>
			<groupId>com.squareup.okhttp3</groupId>
			<artifactId>okhttp</artifactId>
			<version>4.7.2</version>
		</dependency>
		<dependency>
			<groupId>com.squareup.okhttp3</groupId>
			<artifactId>logging-interceptor</artifactId>
			<version>4.7.2</version>
		</dependency>

I thing it would be a good idea to add this in the documentation as springboot is used by many people,also I know that use the basic auth to get the token probably is not the most typica use case.

Anyway I hope this helps someone with same problem
